### PR TITLE
Updated Underscore to handle Spans with same boundaries

### PR DESCRIPTION
**Pull request by nrodnova**
Updated Underscore class to properly handle extension values for `Spans` with equal boundaries. 

Resolves issue https://github.com/explosion/spaCy/issues/9706

## Description
The change actually looks reasonably neat. Nothing changed for `Doc` or `Token`. The only thing that needed to be taken care of was the case when a `Span` that have data in `doc.user_data` had its `label` or `kb_id` updated. I also added a test.

### Types of change

enhancement, new test

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: spacy/tests/doc/test_underscore.py [hunk 1]
```diff
@@ -171,3 +171,59 @@ def test_method(doc, arg1=1, arg2=2):
...
+    doc._.doc_extension = "doc extension"
+    doc[0]._.token_extension = "token extension"
+    span_1._.span_extension = 'span_1 extension'
+    span_2._.span_extension =  'span_2 extension'
+
+    doc._.doc_extension == "doc extension"
+    doc[0]._.token_extension == "token extension"
+    span_1._.span_extension == 'span_1 extension'
+    span_2._.span_extension == 'span_2 extension'
+
```
Comment (svlandeg) [inline spacy/tests/doc/test_underscore.py:196]: It looks like this code was accidentally duplicated.
```suggestion
```

File: spacy/tests/doc/test_underscore.py [hunk 2]
```diff
@@ -171,3 +171,59 @@ def test_method(doc, arg1=1, arg2=2):
...
+    assert doc.user_data[('._.', 'span_extension', span_2.start_char, span_2.end_char, span_2.label, span_2.kb_id)] == 'span_2 extension'
+
+    span_1.label_ = "NEW_LABEL"
+    assert doc.user_data[('._.', 'span_extension', span_1.start_char, span_1.end_char, span_1.label, span_1.kb_id)] == 'span_1 extension'
+    assert doc.user_data[('._.', 'span_extension', span_2.start_char, span_2.end_char, span_2.label, span_2.kb_id)] == 'span_2 extension'
+
+    doc._.doc_extension == "doc extension"
+    doc[0]._.token_extension == "token extension"
+    span_1._.span_extension == 'span_1 extension'
+    span_2._.span_extension == 'span_2 extension'
```
Comment (svlandeg) [inline spacy/tests/doc/test_underscore.py:207]: Is there a reason you copy/pasted this again? I don't think it's necessary?

File: spacy/tests/doc/test_underscore.py [hunk 3]
```diff
@@ -171,3 +171,59 @@ def test_method(doc, arg1=1, arg2=2):
...
+
+    assert doc.user_data[('._.', 'span_extension', span_1.start_char, span_1.end_char, span_1.label, span_1.kb_id)] == 'span_1 extension'
+    assert doc.user_data[('._.', 'span_extension', span_2.start_char, span_2.end_char, span_2.label, span_2.kb_id)] == 'span_2 extension'
+
+    span_1.label_ = "NEW_LABEL"
+    assert doc.user_data[('._.', 'span_extension', span_1.start_char, span_1.end_char, span_1.label, span_1.kb_id)] == 'span_1 extension'
+    assert doc.user_data[('._.', 'span_extension', span_2.start_char, span_2.end_char, span_2.label, span_2.kb_id)] == 'span_2 extension'
```
Comment (svlandeg) [inline spacy/tests/doc/test_underscore.py:202]: I don't think the test for `span_2` should be repeated.
```suggestion
    span_1.label_ = "NEW_LABEL"
    assert doc.user_data[('._.', 'span_extension', span_1.start_char, span_1.end_char, span_1.label, span_1.kb_id)] == 'span_1 extension'
```

File: spacy/tests/doc/test_underscore.py [hunk 4]
```diff
@@ -171,3 +171,59 @@ def test_method(doc, arg1=1, arg2=2):
...
+
+    doc._.doc_extension == "doc extension"
+    doc[0]._.token_extension == "token extension"
+    span_1._.span_extension == 'span_1 extension'
+    span_2._.span_extension == 'span_2 extension'
+
+    span_1.kb_id_ = "KB_ID"
+    assert doc.user_data[('._.', 'span_extension', span_1.start_char, span_1.end_char, span_1.label, span_1.kb_id)] == 'span_1 extension'
+    assert doc.user_data[('._.', 'span_extension', span_2.start_char, span_2.end_char, span_2.label, span_2.kb_id)] == 'span_2 extension'
```
Comment (svlandeg) [inline spacy/tests/doc/test_underscore.py:211]:
```suggestion
```

File: spacy/tests/doc/test_underscore.py [hunk 5]
```diff
@@ -171,3 +171,59 @@ def test_method(doc, arg1=1, arg2=2):
...
+
+    doc._.doc_extension == "doc extension"
+    doc[0]._.token_extension == "token extension"
+    span_1._.span_extension == 'span_1 extension'
+    span_2._.span_extension == 'span_2 extension'
+
```
Comment (svlandeg) [inline spacy/tests/doc/test_underscore.py:217]:
```suggestion
```

File: spacy/tests/doc/test_underscore.py [hunk 6]
```diff
@@ -171,3 +171,59 @@ def test_method(doc, arg1=1, arg2=2):
...
+    assert doc.user_data[('._.', 'doc_extension', None, None)] == 'doc extension'
+    assert doc.user_data[('._.', 'token_extension', 0, None)] == 'token extension'
+
+    span_2._.span_extension = 'updated span_2 extension'
+    assert doc.user_data[('._.', 'span_extension', span_1.start_char, span_1.end_char, span_1.label, span_1.kb_id)] == 'span_1 extension'
```
Comment (svlandeg) [inline spacy/tests/doc/test_underscore.py:222]:
```suggestion
```

File: spacy/tests/doc/test_underscore.py [hunk 7]
```diff
@@ -171,3 +171,59 @@ def test_method(doc, arg1=1, arg2=2):
...
+    assert doc.user_data[('._.', 'span_extension', span_2.start_char, span_2.end_char, span_2.label, span_2.kb_id)] == 'updated span_2 extension'
+
+    doc._.doc_extension == "doc extension"
+    doc[0]._.token_extension == "token extension"
+    span_1._.span_extension == 'span_1 extension'
+    span_2._.span_extension == 'updated span_2 extension'
+
```
Comment (svlandeg) [inline spacy/tests/doc/test_underscore.py:229]:
```suggestion
```

**Review by svlandeg — COMMENTED** [review]
Thanks a lot for creating this PR to make a potential solution for #9706 more tangible! 

I've been reviewing the PR and had some comments about one of the tests, and wanted to check with you whether you could run `black` for autoformatting.

In general, I'm not 100% convinced yet whether this is the way to go. I definitely see the problem you've outlined in the original issue, but I'm wary of this implicit contract between `Span` and `Underscore` that is introduced here. It feels like the code might be prone to bugs at some point, and keeping the labels/keys in-sync might be a challenge.

But I'm not sure what a better solution would be either. Happy to hear feedback from others!

**Comment by svlandeg** [comment]
@adrianeboyd : would love to hear your ideas as well!

**Comment by nrodnova** [comment]
@svlandeg Regarding formatting - I installed pre-commit step for spacy after your comments on another PR. This one was committed before that, that is why there are formatting issues. I will take a look at the tests if you guys decide this idea is a go. I definitely meant to test something particular on every line, but I will double-check.

**Comment by nrodnova** [comment]
@svlandeg Regarding _this implicit contract between Span and Underscore that is introduced here_. We can pass 2 additional arguments explicitly to the `Underscore` constructor from `Doc` and `Token`  (`None` and  `None`), and then there will be no _implicit contract between Span and Underscore_. Current `Underscore` behavior is inconsistent with `Span`'s equality concept. And can cause confusion when `Spans` overwrite each other 's `_` values.

**Comment by svlandeg** [comment]
> Current `Underscore` behavior is inconsistent with `Span`'s equality concept. And can cause confusion when `Spans` overwrite each other 's `_` values.

Yep, I definitely agree that the current behaviour is inconsistent - this API hasn't been updated since we introduced overlapping spans in `doc.spans`.

> @svlandeg Regarding _this implicit contract between Span and Underscore that is introduced here_. We can pass 2 additional arguments explicitly to the `Underscore` constructor from `Doc` and `Token` (`None` and `None`), and then there will be no _implicit contract between Span and Underscore_. 

I was referring to the fact that `Span` needs to know it should update stuff in the underscore data structure whenever a label or KB ID is set. I'm not sure I understand how 2 additional arguments on the constructor would prevent that?

**Comment by adrianeboyd** [comment]
The main difference is modifiable spans. Until `doc.spans`/`spancat` you couldn't edit any properties of a span at all, which did simplify things in a lot of cases, but also made a lot of `spancat` a hassle to implement.

The other issue we need to keep in mind here in serialization. If you create a `DocBin` or run `Doc.to_bytes` with this modified underscore representation, I don't think you can load it with an earlier version of spacy?

**Comment by nrodnova** [comment]
@adrianeboyd I agree that having mutable `Spans` created some hassle, but it is so much more convenient!

Regarding serialization. I would worry more about being able to read binary representations saved with the older model using the newer one, no? Backward compatibility, not the forward one?

**Comment by adrianeboyd** [comment]
It's definitely important that `.spacy` files are both forward- and backward-compatible, at least within spacy v3.x. I added a format version number to the `.spacy` format in v3.0 because I thought it might be useful in the future, but it's not read/checked/used in any way that might help with this problem. I've looked a bit into the formats and readers without seeing an immediate solution...

**Comment by svlandeg** [comment]
Maintenance update: we're putting this specific PR/issue on hold until we've decided on how to proceed with the `shared_ptr` idea: https://github.com/explosion/spaCy/pull/9869

**Comment by adrianeboyd** [comment]
There are also some additional places where this data is referenced or used directly, which would need to be updated after the types of underscore keys are changed. I'm not sure this is everything, but a few places are:

* `Doc.from_docs`
* `Span.as_doc`

**Comment by adrianeboyd** [comment]
Because of the serialization issues, I think this would be best in `v4`, where we've also put the `shared_ptr` modifications.

**Comment by adrianeboyd** [comment]
As a note, this PR should also now add `Span.id` everywhere similar to `Span.kb_id` and `Span.label` so far.

**Comment by svlandeg** [comment]
Closing in favor of #11429 where we're continuing with this work :-)

**Pull request closed without merging**